### ISSUE TITLE
Feature request: cabal sandbox support

### ISSUE THREAD

**Issue by langston-barrett** [xmonad/xmonad#8]
I'd really like to add functionality to xmonad without running into dependency conflicts. There is [a package](https://github.com/Javran/xmonad-entryhelper) that helps with this, perhaps it could be integrated into the core?

**Comment by langston-barrett** [comment]
It looks like [here](https://github.com/xmonad/xmonad/blob/3ce9dcbbb552425f9105ebbf4106d7fb03348f89/src/XMonad/Core.hs#L473-L475) is where Xmonad gets recompiled automatically. We could:
1. introduce a simple `if` to check if there exists a `.cabal` file in the `getXMonadDir` (or alternatively, if `cabal-sandbox.config` exists)
2. execute `cabal build`
3. copy the executable `./dist/build/[project-name]/[binary-name]` to `~/.xmonad/xmonad-x86_64-linux`. 

Would you accept a PR to this effect?

**Comment by seschwar** [comment]
You can easily use XMonad with a Cabal sandbox.  Your `xmonad.hs` doesn't have
to reside in the sandbox, just `xmonad` itself:

```
cd $SOMEWHERE
cabal sandbox init
cabal install xmonad-contrib
cabal exec -- xmonad
cabal exec -- xmonad --restart
```

**Comment by seschwar** [comment]
You can also use XMonad with `stack` in a similar way.

**Comment by langston-barrett** [comment]
I should've been clearer, my apologies. Of course, it is possible to manually compile xmonad in a sandbox. The real feature here would be to do so automatically when available, since xmonad often (re)compiles itself.

**Comment by pjones** [comment]
I believe the best approach for this would be to allow the xmonad user to override the default behavior for rebuilding/restarting.  With that in place I could see new modules in xmonad-contrib for rebuilding with cabal or stack.

Would this be a PR you'd be interested in working on?

**Comment by langston-barrett** [comment]
@pjones It is, I'm not sure I will have the time and knowledge to do so soon, though :(

**Comment by sgf-dma** [comment]
There is a module for defining custom build: https://hackage.haskell.org/package/xmonad-entryhelper .

An example of using it for building with `stack` can be found in my config (well, there're many non-related code, but still..): https://github.com/sgf-dma/sgf-xmonad-config .

May be we can create new interface for overriding building based/similar to this module?

I may try to make a PR, but i don't clearly understand now how interface should look like.

**Comment by geekosaur** [comment]
I was working on an interface for this, but illness (and a bout of second system syndrome) interfered with it. The main issue isn't with this itself, but that there's a number of issues that all touch the same code (proper XDG support being another) and they all end up interacting in such a way that I'd rather try to get them all at once instead of having them break each other. I still need to weed my original thinking down to something sensible though.

**Comment by pjones** [comment]
Fixed by #62.
